hadoop-hdds/rocksdb-checkpoint-differ/src/main/java/org/apache/ozone/rocksdiff/RocksDBCheckpointDiffer.java

```diff
@@ -161,17 +163,23 @@ public class RocksDBCheckpointDiffer implements AutoCloseable {
   private long reconstructionSnapshotGeneration;
   private String reconstructionLastSnapshotID;
 
-  /**
-   * Dummy object that acts as a write lock in compaction listener.
-   */
-  private final Object compactionListenerWriteLock = new Object();
-
   private final ScheduledExecutorService executor;
   private boolean closed;
   private final long maxAllowedTimeInDag;
 
   private ColumnFamilyHandle snapshotInfoTableCFHandle;
 
+  /**
+   * Dummy object that acts as a read and write lock to achieve synchronization
+   * between update to compaction DAG and snapDiff jobs.
+   * Dummy object is used instead of an actual {@link ReadWriteLock} to keep it
+   * simple since all the processes using compaction DAG are background jobs.
+   * They are either RocksDB compaction listener (to add new compaction entries
+   * to DAG), compaction DAG pruning job (to removes older snapshot's from DAG)
+   * or a snap diff job (reads compaction DAG).
+   */
+  private final Object compactionDagLock = new Object();
```

Comment:
We should make RocksDBCheckpointDiffer a singleton class otherwise this could still be a problem with future changes.

Reply:
I don't think it is a good idea to make it singleton in this case because `RocksDBCheckpointDiffer` object is based on [RocksDB's db directory](https://github.com/apache/ozone/blob/fc97e4893c9625864ee32a0e95eba78858f7e60e/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/utils/db/RDBStore.java#L114).

Reply:
Discussed offline. Changing it to singleton will protect from multiple instances get created for `RocksDBCheckpointDiffer`. We want to have only one object of `RocksDBCheckpointDiffer` throughout OM process.

Reply:
After changing RocksDBCheckpointDiffer to simply singleton, most of OM unit and integration tests failed because each test is creating a new instance of OmMetadataManagerImpl which initializes RBDStore and RocksDBCheckpointDiffer. I tired to fix the tests by creating only one instance of OmMetadataManagerImpl per test class but that doesn’t work either. Assertions are failing in the case.

https://github.com/hemantk-12/ozone/actions/runs/4813598710/jobs/8570328203

One way to fix this, is to have one instance of per RocksDBCheckpointDiffer RocksDB dir and keep it in memory. Which solves the unit test failure and is close to what we want achieve.  I made the changes accordingly.